### PR TITLE
feat: add async/await support for `onAuthStateChange` callbacks

### LINKED ISSUE #276: Possibility to sync onAuthStateChange with signIn/Out()
# Feature request

## Is your feature request related to a problem? Please describe.

The code is from a nuxt3 app using the @nuxtjs/supabase module.

The onAuthStateChange listener is often used to the set the user on the client side and its a good and central place to do it imo.

But when having a sign in action like this:

```js
const signIn = async() => {
  const { session, error } = await client.auth.signIn({ email: email.value, password: password.value })
  navigateTo('home')
}
```

with a route guard like this

```js
export default defineNuxtRouteMiddleware((to) => {
  const user = useSupabaseUser()
  if (!(user.value))
    return navigateTo({ name: 'login', query: { redirect: to.path } })
})

```

and a onAuthStateChange listener like this

```js
 // Once Nuxt app is mounted
  nuxtApp.hooks.hook('app:mounted', () => {
    // Listen to Supabase auth changes
    client.auth.onAuthStateChange(async (event: AuthChangeEvent, session: Session | null) => {
      await setServerSession(event, session)
      user.value = client.auth.user()
    })
  })

```

The user is still not available when navigating to the 'home' after login. Causing the middleware to redirect the user back to login.

## Describe the solution you'd like

A possibility in the options to wait for the changeListeners to finish before the login/logout promise resolves.
This would clean up client side code and sync the entire auth process. This could default to false, if it introduces
unwanted behaviour for some cases.

Something like this maybe:

```js
const { session, error } = await client.auth.signIn({ email: email.value, password: password.value },{waitForListeners: true})
```

### PR DESCRIPTION
**Pull request by hf**
If a callback passed to `onAuthStateChange` is `async`, the emitter of the event (like `signInXYZ` methods) would wait before returning. All such callbacks will be waited upon to complete, but errors will be reported once all have completed. All such errors are logged, but only the first one is thrown.

Fixes #276

### CODE REVIEW

File: src/GoTrueClient.ts [hunk 1]
```diff
@@ -1268,12 +1270,33 @@ export default class GoTrueClient {
...
     }
 
-    this.stateChangeEmitters.forEach((x) => x.callback(event, session))
+    const errors: any[] = []
+    const promises = Array.from(this.stateChangeEmitters.values()).map(async (x) => {
+      try {
+        await x.callback(event, session)
+      } catch (e: any) {
+        errors.push(e)
+      }
```
Comment (hf) [inline src/GoTrueClient.ts:1288]: I guess the logging can be done here.

File: src/GoTrueClient.ts [hunk 2]
```diff
@@ -1069,9 +1071,9 @@ export default class GoTrueClient {
       } = await this.getSession()
       if (error) throw error
 
-      this.stateChangeEmitters.get(id)?.callback('INITIAL_SESSION', session)
+      await this.stateChangeEmitters.get(id)?.callback('INITIAL_SESSION', session)
     } catch (err) {
-      this.stateChangeEmitters.get(id)?.callback('INITIAL_SESSION', null)
+      await this.stateChangeEmitters.get(id)?.callback('INITIAL_SESSION', null)
```
Comment (hf) [inline src/GoTrueClient.ts:1076]: I don't think these should be `await`-ed.

Comment (J0) [reply src/GoTrueClient.ts:1076]: bumping this - am keen to know why these statements shouldn't be `await`-ed as well

**Review by J0 — APPROVED** [review]

File: src/GoTrueClient.ts (hunk 2 above)
Comment (hf) [reply src/GoTrueClient.ts:1076]: Trying to remember. 😂

Comment (hf) [reply src/GoTrueClient.ts:1076]: Ah! Nevermind. I thought there was the possibility of recursive await, but doesn't appear so.

**Pull request merged**